tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/archival/InboxArchivalService.java

```diff
@@ -0,0 +1,146 @@
+            .filter(messagesOlderThanArchiveDate(archiveDate))
+            .flatMap(mailboxMessage -> archiveMessage(mailboxMessage, mailboxSession, jmapSettings), ReactorUtils.DEFAULT_CONCURRENCY)
+            .then(Mono.just(Task.Result.COMPLETED))
+            .onErrorResume(e -> {
+                LOGGER.error("Error while archiving INBOX for user {}", username.asString(), e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private FunctionChainer<Mailbox, Flux<MailboxMessage>> getMailboxMessagesMetadata(MailboxSession mailboxSession) {
+        return Throwing.function(mailbox -> mapperFactory.getMessageMapper(mailboxSession)
+            .findInMailboxReactive(mailbox, MessageRange.all(), MessageMapper.FetchType.METADATA, UNLIMITED));
+    }
+
+    private Predicate<MailboxMessage> messagesOlderThanArchiveDate(Date archiveDate) {
+        return mailboxMessage -> mailboxMessage.getSaveDate()
+            .orElse(mailboxMessage.getInternalDate())
+            .before(archiveDate);
+    }
+
+    private Mono<Void> archiveMessage(MailboxMessage mailboxMessage, MailboxSession mailboxSession, JmapSettings jmapSettings) {
+        return switch (jmapSettings.inboxArchivalFormat().toString()) {
+            case "yearly" -> yearlyArchive(mailboxMessage, mailboxSession);
+            case "monthly" -> monthlyArchive(mailboxMessage, mailboxSession);
+            default -> singleArchive(mailboxMessage, mailboxSession);
+        };
+    }
+
+    private Mono<Void> singleArchive(MailboxMessage mailboxMessage, MailboxSession mailboxSession) {
+        return moveMessage(mailboxMessage, mailboxSession, MailboxPath.forUser(mailboxSession.getUser(), DefaultMailboxes.ARCHIVE));
+    }
+
+    private Mono<Void> yearlyArchive(MailboxMessage mailboxMessage, MailboxSession mailboxSession) {
+        int savedYear = mailboxMessage.getSaveDate()
+            .orElse(mailboxMessage.getInternalDate())
+            .toInstant()
+            .atZone(ZoneId.of("UTC"))
+            .toLocalDate()
+            .getYear();
+        MailboxPath yearlyArchiveSubMailbox = MailboxPath.forUser(mailboxSession.getUser(), String.format("Archive.%d", savedYear));
+
+        return moveMessage(mailboxMessage, mailboxSession, yearlyArchiveSubMailbox);
+    }
+
+    private Mono<Void> monthlyArchive(MailboxMessage mailboxMessage, MailboxSession mailboxSession) {
+        LocalDate savedDate = mailboxMessage.getSaveDate()
+            .orElse(mailboxMessage.getInternalDate())
+            .toInstant()
+            .atZone(ZoneId.of("UTC"))
+            .toLocalDate();
+        MailboxPath monthlyArchiveSubMailbox = MailboxPath.forUser(mailboxSession.getUser(),
+            String.format("Archive.%d.%s", savedDate.getYear(), savedDate.getMonth().getDisplayName(TextStyle.FULL, Locale.US)));
+
+        return moveMessage(mailboxMessage, mailboxSession, monthlyArchiveSubMailbox);
```

Comment:
Moving messages one by one is likely sub-optimal. I suggest to group by target folder and try to limit the number of moves.

Reply:
not solved yet

Reply:
Seems not to be able to solve with the current API to me.